### ISSUE TITLE
Reactive context not propagated for participation in Spring-managed reactive transaction

### ISSUE THREAD

**Issue by tomirio619** [jOOQ/jOOQ#17598]
### Expected behavior

Hey!

## TLDR

* jOOQs `DefaultDSLContext.transactionPublisher` does not propagate the Reactor context required for
  participating in a Spring-managed reactive transaction.
* Reproduction: https://gitlab.com/sandsoft/reproduction-repositories/r2dbc-transactional-jooq
    * `InsertRecordUseCaseIT.DSLContext.Should persist no data if transaction-annotated function throws` reproduces
      problem (as a failing test)

## Background

As of Spring Framework version 5.2, reactive transaction management was added [^1].

Using Spring Boot, we can configure jOOQ to use R2DBC (while participating in Spring transactions) as follows:

```kotlin

@Configuration
class JooqConfiguration {
    @Bean
    fun dslContext(connectionFactory: ConnectionFactory): DSLContext {
        return DSL.using(TransactionAwareConnectionFactoryProxy(connectionFactory), SQLDialect.POSTGRES)
    }
}
```

Spring makes transactions work with reactive code by storing metadata in the Reactor/Coroutine context.
If we inspect the `coroutineContext` of a `@Transactional`-annotated `suspend` function, we observe the following
keys in the context:

- `org.springframework.transaction.reactive.TransactionContextHolder`
- `org.springframework.transaction.reactive.TransactionContext`

These context elements are needed by Spring to obtain the `TransactionContext` for reactive code (
see [here](https://github.com/spring-projects/spring-framework/blob/ae279eaced82cf7fe69235556b48befd65290966/spring-tx/src/main/java/org/springframework/transaction/reactive/TransactionContextManager.java#L52-L65)).

## Problem

In our application code, we pass the Coroutine/Reactor context along to jOOQ in the following way:

```kotlin
suspend fun execute(): Unit = coroutineScope {
    dsl.coTransaction(coroutineContext) { scope ->
        scope.dsl()
            .insertInto(table("records"))
            .set(field(name("id"), UUID::class.java), UUID.randomUUID())
            .set(field(name("source"), String::class.java), DSLContext::class.java.simpleName)
            .set(field(name("created_at"), String::class.java), Instant.now().toString())
            .returning()
            .awaitFirstOrNull()
    }
}


suspend fun <T> DSLContext.coTransaction(
    context: CoroutineContext,
    transactional: suspend (Configuration) -> T
): T {
    // [#14997] Wrap values in an auxiliary class in order to allow nulls, which awaitSingle()
    //          doesn't allow, otherwise
    data class Wrap<T>(val t: T)

    return transactionPublisher { c ->
        // Mono context cannot contain a job, so remove it
        val ctx = context.minusKey(Job.Key)
        mono(ctx) {
            Wrap(transactional.invoke(c))
        }
    }.awaitSingle().t
}
```

### Expectation

I expect this code to participate in an existing Spring-managed transaction, but this is not the case.
A transaction rollback does not "undo" the result of the "insert" statement.

### Reproduction

An example project that reproduces this problem can be found
here: https://gitlab.com/sandsoft/reproduction-repositories/r2dbc-transactional-jooq.
The test `InsertRecordUseCaseIT.DSLContext.Should persist no data if transaction-annotated function throws` reproduces
the scenario.
The project also contains the same tests for `DatabaseClient`, which correctly rolls back the transaction.

### Hypothesis

The problem (I believe) is that the call to `transactionPublisher` does not propagate the Reactor/Coroutine context.
As a consequence, Spring is unable to detect that `DSLContext` should participate in a previously-started transaction.
Spring therefore opens a new connection instead.
This detaches the jOOQ-executed SQL statement from the Spring-managed transaction 
(as indicated by the `@Transactional` annotation).

The logs related to the test scenario for `DSLContext` show this:

```
2024-11-11 14:14:03 - o.s.r.core.DefaultDatabaseClient - Executing SQL statement [truncate records]
2024-11-11 14:14:03 - o.s.r.c.R2dbcTransactionManager - Creating new transaction with name [it.sandsoft.reproduction.InsertRecordUseCase.executeJooq]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT
2024-11-11 14:14:03 - o.s.r.c.R2dbcTransactionManager - Acquired Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@56edbcb7, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@7cfa30b}]] for R2DBC transaction
2024-11-11 14:14:03 - o.s.r.c.R2dbcTransactionManager - Starting R2DBC transaction on Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@56edbcb7, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@7cfa30b}]] using [ExtendedTransactionDefinition [transactionName='it.sandsoft.reproduction.InsertRecordUseCase.executeJooq', readOnly=false, isolationLevel=null, lockWaitTimeout=PT0S]]
[DSLContext] - insert data
[DSLContext] - finish insert data  // HERE <--- no detection of existing transaction
2024-11-11 14:14:03 - o.s.r.c.R2dbcTransactionManager - Initiating transaction rollback
2024-11-11 14:14:03 - o.s.r.c.R2dbcTransactionManager - Rolling back R2DBC transaction on Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@56edbcb7, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@7cfa30b}]]
2024-11-11 14:14:03 - o.s.r.c.R2dbcTransactionManager - Releasing R2DBC Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@56edbcb7, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@7cfa30b}]] after transaction
java.lang.RuntimeException: Oeps, something went wrong!
2024-11-11 14:14:03 - o.s.r.core.DefaultDatabaseClient - Executing SQL statement [select * from records]
```

The logs for the test scenario for `DatabaseClient` show how the Spring-managed transaction is detected by `R2dbcTransactionManager`:

```raw
2024-11-11 14:15:42 - o.s.r.core.DefaultDatabaseClient - Executing SQL statement [truncate records]
2024-11-11 14:15:42 - o.s.r.c.R2dbcTransactionManager - Creating new transaction with name [it.sandsoft.reproduction.InsertRecordUseCase.executeDatabaseClient]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT
2024-11-11 14:15:42 - o.s.r.c.R2dbcTransactionManager - Acquired Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@788fe5d1, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@5a91465f}]] for R2DBC transaction
2024-11-11 14:15:42 - o.s.r.c.R2dbcTransactionManager - Starting R2DBC transaction on Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@788fe5d1, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@5a91465f}]] using [ExtendedTransactionDefinition [transactionName='it.sandsoft.reproduction.InsertRecordUseCase.executeDatabaseClient', readOnly=false, isolationLevel=null, lockWaitTimeout=PT0S]]
[DatabaseClient] - insert data
2024-11-11 14:15:42 - o.s.r.c.R2dbcTransactionManager - Participating in existing transaction <--- HERE - detection of existing transaction
2024-11-11 14:15:42 - o.s.r.core.DefaultDatabaseClient - Executing SQL statement [insert into records (id, source, created_at) values (:id, :source, :createdAt)]
[DatabaseClient] - finish insert data
2024-11-11 14:15:42 - o.s.r.c.R2dbcTransactionManager - Initiating transaction rollback
2024-11-11 14:15:42 - o.s.r.c.R2dbcTransactionManager - Rolling back R2DBC transaction on Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@788fe5d1, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@5a91465f}]]
2024-11-11 14:15:42 - o.s.r.c.R2dbcTransactionManager - Releasing R2DBC Connection [PooledConnection[PostgresqlConnection{client=io.r2dbc.postgresql.client.ReactorNettyClient@788fe5d1, codecs=io.r2dbc.postgresql.codec.DefaultCodecs@5a91465f}]] after transaction
java.lang.RuntimeException: Oeps, something went wrong!
2024-11-11 14:15:42 - o.s.r.core.DefaultDatabaseClient - Executing SQL statement [select * from records]
```

### Details

In a bit more detail:

- The implementation of `transactionPublisher` obtains the (transaction-aware) connection factory, and creates a
  `TransactionSubscription` [^1].
- The `TransactionSubscription` [^2] makes the `super(ctx.configuration(), subscriber);` call [^3] and obtains a
  connection from the `Configuration` (using the `ConnectionFactory`) [^4].
- Eventually, `ConnectionFactoryUtils.doGetConnection(ConnectionFactory connectionFactory)` is called [^5].
- `TransactionSynchronizationManager.forCurrentTransaction()` tries to obtain the `TransactionSynchronizationManager`
  that is associated with the current (Reactor) transaction context [^6] [^7].
- This logic is unable to detect the existing transaction as the context is empty.

Spring's `DatabaseClient` is correctly participating in the already-opened transaction due to Spring awareness of that transaction due to context propagation.

[^1]: https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-5.2-Release-Notes#transaction-management
[^2]: https://github.com/jOOQ/jOOQ/blob/6195a73b62ba76c8a4344f8f73b740c0b4a9c1bc/jOOQ/src/main/java/org/jooq/impl/DefaultDSLContext.java#L630-L640
[^3]: https://github.com/jOOQ/jOOQ/blob/cb5e14e601da8340aa77d34c402f98b2f446e0d7/jOOQ/src/main/java/org/jooq/impl/R2DBC.java#L772
[^4]: https://github.com/jOOQ/jOOQ/blob/cb5e14e601da8340aa77d34c402f98b2f446e0d7/jOOQ/src/main/java/org/jooq/impl/R2DBC.java#L781
[^5]: https://github.com/jOOQ/jOOQ/blob/cb5e14e601da8340aa77d34c402f98b2f446e0d7/jOOQ/src/main/java/org/jooq/impl/R2DBC.java#L624
[^6]: https://github.com/spring-projects/spring-framework/blob/e778d2e908d36c5062c6124179a31f5ccdabbe78/spring-r2dbc/src/main/java/org/springframework/r2dbc/connection/ConnectionFactoryUtils.java#L112
[^7]: https://github.com/spring-projects/spring-framework/blob/6baa60d4548b0b2a738c3be7266232f9c8f8a0a3/spring-tx/src/main/java/org/springframework/transaction/reactive/TransactionSynchronizationManager.java#L89
[^8]: https://github.com/spring-projects/spring-framework/blob/ae279eaced82cf7fe69235556b48befd65290966/spring-tx/src/main/java/org/springframework/transaction/reactive/TransactionContextManager.java#L52-L65

### Actual behavior

see above

### Steps to reproduce the problem

see above

### jOOQ Version

jOOQ Open Source 3.19.15

### Database product and version

postgreSQL 15

### Java Version

openjdk 21.0.1 2023-10-17 LTS

### JDBC / R2DBC driver name and version (include name if unofficial driver)

org.postgresql:r2dbc-postgresql:1.0.7.RELEASE

**Comment by lukaseder** [comment]
Thanks for your report. I think this is a duplicate of the following issue (to be confirmed):

- https://github.com/jOOQ/jOOQ/issues/16109

In short, jOOQ cannot interact with Reactor because it doesn't know about reactor, only reactive streams and r2dbc, both of which don't have any such thing as a "context." This has been discussed but doesn't look like it's on the roadmap for the RS SPI anytime soon:

- https://github.com/reactive-streams/reactive-streams-jvm/issues/542

I don't think jOOQ can do anything about this?

**Comment by tomirio619** [comment]
Ah yes that seems to be the exact same problem I'm running into here.
I hoped it would be a minor change on the application side to get this working, but given that jOOQ is not depending (and won't?) on reactor, I don't see a nice way to do it.
Maybe good to document this somewhere in the docs?

**Comment by tomirio619** [comment]
As mentioned here (https://github.com/jOOQ/jOOQ/issues/16109#issuecomment-2461831483), something like this would to the trick:

```kotlin
suspend fun execute() {
    dsl.coTransaction(connectionFactory) {
        it.insertInto(table("records"))
            .set(field(name("id"), UUID::class.java), UUID.randomUUID())
            .set(field(name("source"), String::class.java), DSLContext::class.java.simpleName)
            .set(field(name("created_at"), String::class.java), Instant.now().toString())
            .returning()
            .awaitFirstOrNull()
    }
}

suspend fun <T> DSLContext.coTransaction(
    connectionFactory: ConnectionFactory,
    transactional: suspend (DSLContext) -> T
): T {
    val connection = ConnectionFactoryUtils.getConnection(connectionFactory).awaitSingle()
    return transactional.invoke(
        using(connection, SQLDialect.POSTGRES, this.settings()).dsl()
    )
}
```

**Comment by lukaseder** [comment]
> Maybe good to document this somewhere in the docs?

I'm not sure why it should be documented in the jOOQ docs. It has nothing to do with jOOQ or with reactive streams or with R2DBC. It seems to be a reactor/spring feature. jOOQ's docs can't be filled with hundreds of integration topics, because that would just keep us busy keeping unrelated integration topics up to date. That's also why all the examples, including outdated Spring examples were recently deleted. Imagine, for a library vendor, having to find out all the hundreds of flaws of various integrators, like Spring, Quarkus, etc. etc. And also to keep all those things up to date. It's a nightmare. :)

The best way to document this, in my opinion, is to ask and answer a question on Stack Overflow, for example:
https://stackoverflow.com/help/self-answer

The second best way is to ask the Spring Boot teams to document this limitation, because their starters mindset suggests that stuff "just works," but it doesn't, in this case.

> As mentioned here ([#16109 (comment)](https://github.com/jOOQ/jOOQ/issues/16109#issuecomment-2461831483)), something like this would to the trick:

Perhaps this is worth suggesting to Spring Boot, to add to their starters?

**Comment by lukaseder** [comment]
I'm closing this as a duplicate of https://github.com/jOOQ/jOOQ/issues/16109

**Comment by lukaseder** [comment]
jOOQ 3.20 will ship with a new SPI to allow for propagating third party contexts through jOOQ. This hsould resolve the issue here:

- https://github.com/jOOQ/jOOQ/issues/17920
